Uploader/TraceFileScanner.cs

```diff
@@ -0,0 +1,147 @@
+﻿using NLog;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+/// <summary>
+/// Scans the trace directory for trace files that are ready to upload or archive.
+/// </summary>
+public class TraceFileScanner
+{
+    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+    private static readonly Regex TraceFileRegex = new Regex(@"^coverage_\d*_\d*.txt$");
+
+    private readonly string traceDirectory;
+    private readonly Regex versionAssemblyRegex;
+    private readonly IFileSystem fileSystem;
+
+    public TraceFileScanner(string traceDirectory, string versionAssembly, IFileSystem fileSystem)
+    {
+        this.traceDirectory = traceDirectory;
+        this.versionAssemblyRegex = new Regex(@"^Assembly=" + Regex.Escape(versionAssembly) + @".*Version:([^ ]*).*", RegexOptions.IgnoreCase);
+        this.fileSystem = fileSystem;
+    }
+
+    /// <summary>
+    /// Returns all trace files that can be uploaded or archived.
+    /// </summary>
+    public IEnumerable<ScannedFile> ListTraceFilesReadyForUpload()
+    {
+        List<string> files;
+        try
+        {
+            files = fileSystem.Directory.EnumerateFiles(traceDirectory).ToList();
```

Comment:
why not use foreach for enumeration? EnumerateFiles(traceDirectory).Where(IsTraceFile)